File: vuln-sandbox/node/taint/path-2hop.js
// Taint: Path Traversal — 2-hop (JS)
// Source: req.params.name → intermediate: buildPath() → Sink: fs.readFileSync
// Expected: TruePositive (taint/cwe22)
const express = require('express');
const fs = require('fs');
const app = express();

function buildPath(name) {
    return fs.readFileSync(`/data/${name}`);    // SINK: file read
}

app.get('/file/:name', (req, res) => {
    const name = req.params.name;              // SOURCE: HTTP request param
    const content = buildPath(name);           // INTERMEDIATE
    res.send(content);
});

File: vuln-sandbox/python/taint/path-2hop.py
# Taint: Path Traversal — 2-hop (Python)
# Source: request.args.get('name') → intermediate: read_file() → Sink: open()
# Expected: TruePositive (taint/cwe22)
from flask import Flask, request

app = Flask(__name__)

def read_file(name):
    with open(f"/data/{name}") as f:           # SINK: file open
        return f.read()

@app.route('/file')
def get_file():
    name = request.args.get('name')            # SOURCE: HTTP request param
    return read_file(name)                     # INTERMEDIATE
